keps/sig-storage/20191012-storagepool-advplacement.md

```diff
@@ -0,0 +1,364 @@
+
+The placement choices in this use case cannot be expressed using existing Kubernetes mechanism.
+
+![Story2-LocalDrives](./kep-storagepool-story2.png)
+
+Beyond the existence of the Placement Choice / Storage Pool, Minio further uses metadata to know that the storage pool has enough capacity to accommodate the desired PVC. If for example physical disks have different size, then Minio may decide to create PVCs of different sizes to still work in this scenario.
+
+#### Story 3
+
+Let’s assume a sharded and replicated MongoDB instance. Let’s further assume that the Cloud Storage Service used for the PVCs is comprised of 3 distinct Storage-Pods (not to be confused with K8s Pods) in both of the two available AZs, such that the 3 Storage-Pods fail independently by having dedicated power, networking, racks, hosts and disks. The CSI driver of the Cloud Storage service announces that 3 Storage-Pods per AZ as placement choices (AZ1-SP1, AZ1-SP2, AZ1-SP3, AZ2-SP1, etc.) and also advertises that all nodes in AZ1 can access all Storage-Pods in AZ1 (as a simple example). Let’s further assume MongoDB is configured for 3-way replication per shard. MongoDB decides to create 3 Pods for a given shard:
+* Pod1 has an affinity to AZ1, and its PVCs have an affinity to AZ1-SP1.
+* Pod2 has an affinity to AZ1, and its PVCs have an affinity to AZ1-SP2.
+* Pod3 has an affinity to AZ2, and its PVCs have an affinity to AZ2-SP1.
+
+If we compare this to the case where we can’t express the PVC affinity to the Storage-Pod, Pod1 and Pod2 may land on the same Storage-Pod and so if that Storage-Pod fails, it would take down two Pods. In the scenario described here, the math will show a higher availability and durability.
+
+The placement choices in this use case cannot be expressed using existing Kubernetes mechanism.
+
+![Story3-MultiStoragePods](./kep-storagepool-story3.png)
+
+
+### Implementation Details/Notes/Constraints
+
+#### New CRD to surface Storage Pools
+```
+apiVersion: storagepool.storage.k8s.io/v1alpha1
+kind: StoragePool
+metadata:
+  name: storagePool1
+  labels:
+     labelKey: labelValue
+spec:
+  driver: myDriver
+  parameters:
+     csiBackendSpecificOpaqueKey: csiBackendSpecificOpaqueValue
+status:
+  accessibleNodes: 
+  - node1
+  - node2
+  capacity:
+    total: 124676572
+```
```

Comment:
I think this CRD can be installed in the similar way as the snapshot beta CRD.  It can be installed by the cluster addon manager.

Reply:
Do you have some pointers for me about that? Are there language bindings involved and how does Kubernetes pull those in?

Reply:
We discussed this a bit on the #csi Slack channel:
- the snapshot beta CRD is not used by core Kubernetes, so it's not an example that applies here
- the addon manager that @xing-yang referred to is https://github.com/kubernetes/kubernetes/tree/master/cluster/addons, which is deprecated and not a general solution for installing a CRD

To me it still looks like core API extensions need to be done in apiserver.

Reply:
The reason that CSINode was transitioned from CRDs to in-tree APIs was mainly because kubelet depends on it.

Unless if we can prove that we can's use CRDs to achieve our goals, we can't explain to sig-architecture why StoragePool has to be in-tree APIs.

Reply:
> The reason that CSINode was transitioned from CRDs to in-tree APIs was mainly because kubelet depends on it.

The same applies here once we want to use the capacity information also for scheduling pods because then the Kubernetes controller depends on the capacity API - see https://github.com/kubernetes/enhancements/pull/1353.